app/helpers/lookup_keys_helper.rb

```diff
@@ -68,39 +68,122 @@ def validator_type_selector(f)
+  def overridable_lookup_keys(klass, obj)
+    klass.class_params.override.where(:environment_classes => {:environment_id => obj.environment}) + klass.lookup_keys
   end
 
-  def hostgroup_key_with_diagnostic(hostgroup, key)
-    value, origin = hostgroup.inherited_lookup_value key
-    original_value = key.value_before_type_cast value
-    diagnostic_helper = popover('', _("<b>Description:</b> %{desc}<br><b>Type:</b> %{type}<br> <b>Matcher:</b> %{matcher}") % { :desc => key.description, :type => key.key_type, :matcher => origin}, :data => { :placement => 'top' })
-    parameter_value_content("value_#{key.key}", original_value, :popover => diagnostic_helper)
+  def can_edit_params?
+    authorized_via_my_scope("host_editing", "edit_params")
   end
 
-  def host_key_with_diagnostic(host, value_hash, key)
-    value_for_key = value_hash[key.id] && value_hash[key.id][key.key]
-    value, matcher = value_for_key ? [value_for_key[:value], "#{value_for_key[:element]} (#{value_for_key[:element_name]})"] : [key.default_value, _("Default value")]
-    original_value = key.value_before_type_cast value
-    no_value = value.nil? && key.lookup_values.find_by_match("fqdn=#{host.fqdn}")
+  def lookup_key_with_diagnostic(obj, lookup_key, lookup_value)
+    value, matcher = value_matcher(obj, lookup_key)
+    inherited_value = lookup_key.value_before_type_cast(value)
+    overridden  = inherited_value.present? || lookup_value.value.present?
+    warnings  = lookup_key_warnings(lookup_key.required, overridden)
 
-    if no_value
-      if key.required
-        diagnostic_class = 'error'
-        diagnostic_helper = popover('', _("Required parameter without value.<br/><b>Please override!</b> <br><br><b>Description:</b>: %s") % key.description, :icon => "exclamation-sign")
+    parameter_value_content(
+      "#{parameters_receiver}_lookup_values_attributes_#{lookup_key.id}_value",
+      lookup_value.value || inherited_value,
+      :popover => diagnostic_popover(lookup_key, matcher, inherited_value, warnings),
+      :name => "#{lookup_value_name_prefix(lookup_key.id)}[value]",
+      :disabled => !lookup_key.overridden?(obj) || lookup_value.use_puppet_default,
+      :inherited_value => inherited_value)
+  end
+
+  def value_matcher(obj, lookup_key)
+    if obj.class.model_name == "Host"
+      value = value_hash_cache(obj)[lookup_key.id]
+      value_for_key = value.try(:[], lookup_key.key)
+      if value_for_key.present?
+        [value_for_key[:value],
+          "#{value_for_key[:element]} (#{value_for_key[:element_name]})"]
       else
-        diagnostic_class = 'warning'
-        diagnostic_helper = popover('', _("Optional parameter without value.<br/><i>Won\'t be given to Puppet.</i> <br><br><b>Description:</b> %s") % key.description, :icon => "warning-sign")
+          [lookup_key.default_value, _("Default value")]
       end
+    else #hostgroup
+      obj.inherited_lookup_value(lookup_key)
```

Comment:
How about moving these to their own helpers, or the model even? 
Ideally we should have a HostForm and HostgroupForm, or  call `value_matcher(obj, lookup_key)` and it should respond from the proper helper. The only reason why that doesn't work is because we load *all* helpers on *all* views and there's a name conflict. 

But moving this logic to the model itself would be fine and respond well. You would just call `obj.value_matcher(lookup_key)` where needed and it responds properly depending on what `obj` is, without type checking.